### PR TITLE
Order Details > variation details: storage layer changes

### PR DESCRIPTION
**Pull request by jaclync**
Storage layer changes for #2280

This PR only contains changes in the storage layer, since the Networking/Yosemite changes are going to be huge as well 🙂 

## Changes

- Schema updates (please review the diffs on model 35 in https://github.com/woocommerce/woocommerce-ios/commit/2c63d3fcef2155e2fdda64909343faa6506c275c#diff-2f9c7ea9bf9793125e3c162b4fcb6dbd5ff9c50054f30aacfc2b996b9c11de71)
  - Added `OrderItemAttribute` entity
  - Added  `attributes: [OrderItemAttribute]` relationship to `OrderItem`
- Added a unit test for migrating from mode 34 to 35 in `MigrationTests`

## Testing

Please run a confidence check on migration!

- Launch the app from `develop` and log in or stay logged in
- Stop the app
- Launch the app from PR branch --> the app should launch in logged in state with the same data

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Storage/Storage/Model/MIGRATIONS.md
Comment (pmusolino) [inline Storage/Storage/Model/MIGRATIONS.md]: Probably it's better to merge this PR in release `5.5.0.0`?

File: Storage/Storage/Model/OrderItemAttribute+CoreDataProperties.swift
Comment (pmusolino) [inline Storage/Storage/Model/OrderItemAttribute+CoreDataProperties.swift:16]: Do you think that it worth adding the optional string field `type` that [will be introduced in the future](https://a8c.slack.com/archives/CGPNUU63E/p1603926767142700?thread_ts=1602796453.088400&cid=CGPNUU63E)?

**Review by pmusolino — APPROVED** [review]
Thanks, @jaclync for introducing the new entity.

I've left some comments and I've two questions:
- Can I ask you to add the `Model 35` also inside the test method `test_it_can_load_the_expected_model_versions` under `ManagedObjectModelsInventoryTests` please?
- An `OrderItemAttribute` can be used by other Order Items or they are unique? If they can be reused by other Order Items, probably it makes sense to change the Delete Rule of `attributes` inside `OrderItem` from `Cascade` to `Nullify`.

File: Storage/Storage/Model/OrderItemAttribute+CoreDataProperties.swift
Comment (jaclync) [reply Storage/Storage/Model/OrderItemAttribute+CoreDataProperties.swift:16]: Hmm I was thinking we can add this when the value is available in the API (the migration doesn't require a mapping model, we can set its default to empty string) - @shiki what do you think, since you're thinking about adding `type`?

File: Storage/Storage/Model/MIGRATIONS.md
Comment (jaclync) [reply Storage/Storage/Model/MIGRATIONS.md]: Hah yea, let's start following the no-merge-big-changes-on-Thursday/Friday-before-code-freeze guideline! I'll merge it after code freeze on my Tuesday next week

**Comment by jaclync** [comment]
> - Can I ask you to add the `Model 35` also inside the test method `test_it_can_load_the_expected_model_versions` under `ManagedObjectModelsInventoryTests` please?

interesting, I didn't see this item in our [Core Data migration documentation](https://github.com/woocommerce/woocommerce-ios/blob/develop/docs/creating-core-data-model-versions.md) and the test case `test_it_can_load_the_expected_model_versions` intentionally not fails on missing latest model versions. @shiki do you think it makes sense to just enforce the complete model version names in the test case? like removing the truncation here:

https://github.com/woocommerce/woocommerce-ios/blob/bb6e0a7defc4ba99e1955f2060e2c35e63e9aeec/Storage/StorageTests/CoreData/ManagedObjectModelsInventoryTests.swift#L74-L77

just wondering if you have other concerns since you created the test case this way :) @pmusolino I also added "Model 35" in https://github.com/woocommerce/woocommerce-ios/pull/3078/commits/4a3692151dbbb5d04553ae8712030170f0b7f0d1

> - An `OrderItemAttribute` can be used by other Order Items or they are unique? If they can be reused by other Order Items, probably it makes sense to change the Delete Rule of `attributes` inside `OrderItem` from `Cascade` to `Nullify`.

I don't think an `OrderItemAttribute` is reused by more than one `OrderItem`, since an attribute is specific to an order item. I think it's easier to maintain this way, and will be more obvious in the next PR with Yosemite/Networking changes

File: Storage/Storage/Model/MIGRATIONS.md
Comment (jaclync) [reply Storage/Storage/Model/MIGRATIONS.md]: updated in https://github.com/woocommerce/woocommerce-ios/pull/3078/commits/26050612ded63832d6dd6d1824a3869a8d34e5c9

File: Storage/Storage/Model/OrderItemAttribute+CoreDataProperties.swift
Comment (shiki) [reply Storage/Storage/Model/OrderItemAttribute+CoreDataProperties.swift:16]: Let's probably wait because that will have to through a code review. Maybe the Core team will not like it and suggest something different. 🙂

**Comment by jaclync** [comment]
Merging the Core Data schema changes now that `develop` is 5.5!

**Comment by shiki** [comment]
> @shiki do you think it makes sense to just enforce the complete model version names in the test case? like removing the truncation here:

@jaclync Sorry I missed this. 

I never intended for that test to fully capture all the models. That test is really just proving that it loads model versions in sequence. It's a specifically scoped unit test for the `ManagedObjectModelsInventory` struct itself and not for the model versions. 

It doesn't need to be updated but it doesn't hurt to update it. 🙂  I don't believe we should enforce it though. It's an unnecessary addition to the process that doesn't provide a lot of benefits in my opinion.

**Pull request merged**
